freud/diffraction.pyx

```diff
@@ -0,0 +1,316 @@
+    :math:`z` axis (at the :math:`xy` plane). The points in the system are
+    converted to fractional coordinates, then binned into a grid whose
+    resolution is given by ``grid_size``. A higher ``grid_size`` will lead to
+    a higher resolution. The points are convolved with a Gaussian of width
+    :math:`\sigma`, given by ``peak_width``. This convolution is performed
+    as a multiplication in Fourier space. The computed diffraction pattern
+    is returned as a square array of shape ``(output_size, output_size)`.
+
+    Args:
+        grid_size (unsigned int):
+            Resolution of the diffraction grid. (Default value = 512).
+        output_size (unsigned int):
+            Size of the output diffraction image (Default value = 512).
+        zoom (float):
+            Scaling factor for incident wavevectors (Default value = 4).
+        peak_width (float):
+            Width of Gaussian convolved with points, in system length units
+            (Default value = 1).
+    """
+    cdef int grid_size
+    cdef int output_size
+    cdef double zoom
+    cdef double peak_width
+    cdef double[:] _k_values_orig
+    cdef double[:, :, :] _k_vectors_orig
+    cdef double[:] _k_values
+    cdef double[:, :, :] _k_vectors
+    cdef double[:, :] _diffraction
+
+    def __init__(self, grid_size=512, output_size=512, zoom=4, peak_width=1):
+        self.grid_size = int(grid_size)
+        self.output_size = int(output_size)
+        self.zoom = np.double(zoom)
+        self.peak_width = np.double(peak_width)
+
+        # Cache these because they are system-independent.
+        self._k_values_orig = np.empty(grid_size)
+        self._k_vectors_orig = np.empty((grid_size, grid_size, 3))
+
+        # Store these computed arrays which are exposed as properties.
+        self._k_values = np.empty_like(self._k_values_orig)
+        self._k_vectors = np.empty_like(self._k_vectors_orig)
+        self._diffraction = np.empty((grid_size, grid_size))
+
+    def _calc_proj(self, view_orientation, box):
+        """Calculate the inverse shear matrix from finding the projected box
+        vectors whose area of parallogram is the largest.
+
+        Args:
+            view_orientation ((:math:`4`) :class:`numpy.ndarray`):
+                View orientation.
+            box (:class:`~.box.Box`):
+                Simulation box.
+
+        Returns:
+            (2, 2) :class:`numpy.ndarray`:
+                Inverse shear matrix.
+        """
+        # Rotate the box matrix by the view orientation.
+        box_matrix = rowan.rotate(view_orientation, box.to_matrix())
+
+        # Compute normals for each box face.
+        # The area of the face is the length of the vector.
+        box_face_normals = np.cross(
+            np.roll(box_matrix, 1, axis=-1),
+            np.roll(box_matrix, -1, axis=-1),
+            axis=0)
+
+        # Compute view axis projections.
+        projections = np.abs(box_face_normals.T @ np.array([0., 0., 1.]))
```

Comment:
This is not a change request, but a question specifically related to @bdice guidance on a PR I made -- is my understanding of this correct? Matrix operations are okay here because they maintain double float precision, while calculations written with single precision floats should be avoided due to potential rounding issues?

Reply:
It's possible that `box_face_normals` is in single (`float`) precision. I'm not sure, but ultimately it doesn't matter because it will be upcast to `double` precision somewhere along the way. The results of math operations are always upcast:
```python
>>> import numpy as np
>>> result = np.float(3.5) * np.double(7.2)
>>> type(result)  # This is a double because of upcasting
<class 'numpy.float64'>
```
If we were using something like Cython's `cdef float[:]` for data arrays, then it would be cast to single-precision when it is stored in such a variable.